File: algorithm/[book]c_algorithm/src/f.6-2_move_to_front/MoveToFront.c
/*
 * MoveToFront.c
 *
 *  Created on: 2019. 1. 5.
 *      Author: otrodevym
 */
#include "MoveToFront.h"

Node* SLL_MoveToFront(Node** Head, int Target){
	Node* Current = (*Head);
	Node* Previous = NULL;
	Node* Match = NULL;


	while(Current != NULL){
		if(Current->Data == Target){
			Match = Current;
			if(Previous != NULL){
				Previous->NextNode = Current->NextNode;
				Current->NextNode = (*Head);
				(*Head) = Current;
			}
				break;

		}else{
			Previous = Current;
			Current = Current->NextNode;
		}
	}
	return Match;
}

File: algorithm/[book]Java_Algorithm/src/f_search/MoveToFront.java
package f_search;
public class MoveToFront {
	private Node head;
	private int size;
	
	public MoveToFront() {
	}
	
	
	private void add(int value) {
		Node newNode = new Node(value);
		if(this.head == null) {
			this.head = newNode;
		}else {
			Node temp = this.head;
			while(temp.getNextNode() != null) {
				temp = temp.getNextNode();
			}
			
			temp.setNextNode(newNode);
		}
	}
	
	
	private Node MoveToFront(int target) {
		Node current = this.head;
		Node previous = null;
		Node match = null;
		
		while(current != null) {
			if(current.getValue() == target) {
				match = current;
				if(previous != null) {
					previous.setNextNode(current.getNextNode());
					current.setNextNode(this.head);
					this.head = current;
				}
				break;
			}else {
				previous = current;
				current = current.getNextNode();
				
			}
		}
		
		return match;
	}
	
	private void printTree() {
		Node temp = this.head;
		while(temp.getNextNode()!= null) {
			System.out.print(" " + temp.getValue());
			temp = temp.getNextNode();
		}
	}
	public static void main(String args[]) {
		MoveToFront mtf = new MoveToFront();
		mtf.add(1);
		mtf.add(2);
		mtf.add(3);
		mtf.add(4);
		mtf.add(5);
		mtf.add(6);
		
		mtf.MoveToFront(4);
		
		mtf.printTree();
		
	}
}